pennylane/measure.py

```diff
@@ -15,11 +15,190 @@
+from pennylane.wires import Wires
+
+
+class MeasurementProcess:
+    """Represents a measurement process occurring at the end of a
+    quantum variational circuit.
+
+    Args:
+        return_type (.ObservableReturnTypes): The type of measurement process.
+            This includes ``Expectation``, ``Variance``, ``Sample``, ``State``, or ``Probability``.
+        obs (.Observable): The observable that is to be measured as part of the
+            measurement process. Not all measurement processes require observables (for
+            example ``Probability``); this argument is optional.
+        wires (.Wires): The wires the measurement process applies to.
+            This can only be specified if an observable was not provided.
+        eigvals (array): A flat array representing the eigenvalues of the measurement.
+            This can only be specified if an observable was not provided.
+    """
+
+    # pylint: disable=too-few-public-methods
+
+    def __init__(self, return_type, obs=None, wires=None, eigvals=None):
+        self.return_type = return_type
+        self.obs = obs
+
+        if wires is not None and obs is not None:
+            raise ValueError("Cannot set the wires if an observable is provided.")
+
+        self._wires = wires or Wires([])
+        self._eigvals = None
+
+        if eigvals is not None:
+            if obs is not None:
+                raise ValueError("Cannot set the eigenvalues if an observable is provided.")
+
+            self._eigvals = np.array(eigvals)
+
+        # TODO: remove the following lines once devices
+        # have been refactored to accept and understand recieving
+        # measurement processes rather than specific observables.
+
+        # The following lines are only applicable for measurement processes
+        # that do not have corresponding observables (e.g., Probability). We use
+        # them to 'trick' the device into thinking it has recieved an observable.
+
+        # Below, we imitate an identity observable, so that the
+        # device undertakes no action upon recieving this observable.
+        self.name = "Identity"
+        self.data = []
+
+        # Queue the measurement process
+        self.queue()
+
+    def diagonalizing_gates(self):
+        """Returns the gates that diagonalize the measured wires such that they
+        are in the eigenbasis of the circuit observables.
+
+        Returns:
+            List[.Operation]: the operations that diagonalize the observables
+        """
+        try:
+            return self.expand().operations
+        except NotImplementedError:
+            return []
+
+    def __repr__(self):
+        """Representation of this class."""
+        if self.obs is None:
+            return "{}(wires={})".format(self.return_type.value, self.wires.tolist())
+
+        # Todo: when tape is core the return type will always be taken from the MeasurementProcess
```

Comment:
Is this TODO done or does something need to updated here?
```suggestion
```

Reply:
Oh, good catch! @mariaschuld I think you wrote this todo? I have to admit, I'm not sure what I need to change here.